packages/core/src/molecule/codec.advanced.ts

```diff
@@ -0,0 +1,373 @@
+ * @param itemCodec the vector item codec. It can be fixed-size or dynamic-size.
+ */
+export function dynItemVec<Encodable>(
+  itemCodec: Codec<Encodable>,
+): Codec<Array<Encodable>> {
+  return bytes({
+    encode(userDefinedItems) {
+      const encodedArray = userDefinedItems.map((item) =>
+        itemCodec.encode(item),
+      );
+      const packed = encodedArray.reduce(
+        (result, item) => {
+          const packedHeader = Uint32LE.encode(BigInt(result.offset));
+          return {
+            header: bytesConcat(result.header, packedHeader),
+            body: bytesConcat(result.body, item),
+            offset: result.offset + bytesFrom(item).byteLength,
+          };
+        },
+        {
+          header: new ArrayBuffer(0),
+          body: new ArrayBuffer(0),
+          offset: 4 + userDefinedItems.length * 4,
+        },
+      );
+      const packedTotalSize = Uint32LE.encode(
+        BigInt(packed.header.byteLength + packed.body.byteLength + 4),
+      );
+      return bytesConcat(packedTotalSize, packed.header, packed.body);
+    },
+    decode(buffer) {
+      const value = bytesFrom(buffer);
+      const byteLength = Uint32LE.decode(value.slice(0, 4));
+      if (byteLength !== BigInt(value.byteLength)) {
+        throw new Error(
+          `dynItemVec: invalid buffer size, expected ${byteLength}, but got ${value.byteLength}`,
+        );
+      }
+      const decodedArray: Array<Encodable> = [];
+      if (byteLength <= 4) {
+        return decodedArray;
+      } else {
+        const offset = Number(Uint32LE.decode(value.slice(4, 8)));
+        const itemCount = (offset - 4) / 4;
+        const offsets = new Array(itemCount)
+          .fill(1)
+          .map((_, index) =>
+            Uint32LE.decode(value.slice(4 + index * 4, 8 + index * 4)),
+          );
+        offsets.push(byteLength);
+        for (let index = 0; index < offsets.length - 1; index++) {
+          const start = Number(offsets[index]);
+          const end = Number(offsets[index + 1]);
+          const itemBuffer = value.slice(start, end);
+          decodedArray.push(itemCodec.decode(itemBuffer));
+        }
+        return decodedArray;
+      }
+    },
+  });
+}
+
+/**
+ * General vector codec, if `itemCodec` is fixed size type, it will create a fixvec codec, otherwise a dynvec codec will be created.
+ * @param itemCodec
+ */
+export function vector<Encodable>(
+  itemCodec: Codec<Encodable>,
+): Codec<Array<Encodable>> {
+  if (itemCodec.byteLength !== undefined) {
+    return fixedItemVec(itemCodec);
+  }
+  return dynItemVec(itemCodec);
+}
+
+/**
+ * Option is a dynamic-size type.
+ * Serializing an option depends on whether it is empty or not:
+ * - if it's empty, there is zero bytes (the size is 0).
+ * - if it's not empty, just serialize the inner item (the size is same as the inner item's size).
+ * @param innerCodec
+ */
+export function option<Encodable>(
+  innerCodec: Codec<Encodable>,
+): Codec<Encodable | undefined> {
```

Comment:
`<Encodable | undefined | null, Encodable | undefined>` should be better

Reply:
the packed result won't be undefined, it's just an empty buffer